src/zenml/integrations/hyperai/orchestrators/hyperai_orchestrator.py

```diff
@@ -0,0 +1,445 @@
+        Returns:
+            A `StackValidator` instance.
+        """
+        return StackValidator(
+            required_components={
+                StackComponentType.CONTAINER_REGISTRY,
+                StackComponentType.IMAGE_BUILDER,
+            }
+        )
+
+    def get_orchestrator_run_id(self) -> str:
+        """Returns the active orchestrator run id.
+
+        Raises:
+            RuntimeError: If the environment variable specifying the run id
+                is not set.
+
+        Returns:
+            The orchestrator run id.
+        """
+        try:
+            return os.environ[ENV_ZENML_HYPERAI_RUN_ID]
+        except KeyError:
+            raise RuntimeError(
+                "Unable to read run id from environment variable "
+                f"{ENV_ZENML_HYPERAI_RUN_ID}."
+            )
+
+    def _generate_valid_path_format(self, path) -> str:
+        """Validates if a given string is in a valid path format.
+
+        Args:
+            path (str): The path to be validated.
+
+        Returns:
+            bool: True if the path has a valid format, False otherwise.
+
+        Raises:
+            RuntimeError: If the path is not in a valid format.
+        """
+        # Define a regular expression pattern to match a valid path format
+        pattern = r'^(?:[a-zA-Z]:\\(\\[^\\/:*?"<>|]*)*$|^/([^\0]*)*$)'
+
+        if bool(re.match(pattern, path)):
+            return path
+        else:
+            raise RuntimeError(
+                f"Path '{path}' is not in a valid format, so a mount cannot be established."
+            )
+
+    def prepare_or_run_pipeline(
+        self,
+        deployment: "PipelineDeploymentResponse",
+        stack: "Stack",
+        environment: Dict[str, str],
+    ) -> Any:
+        """Sequentially runs all pipeline steps in Docker containers.
+
+        Assumes that:
+        - A HyperAI (hyperai.ai) instance is running on the configured IP address.
+        - The HyperAI instance has been configured to allow SSH connections from the
+            machine running the pipeline.
+        - Docker and Docker Compose are installed on the HyperAI instance.
+        - A key pair has been generated and the public key has been added to the
+            HyperAI instance's `authorized_keys` file.
+        - The private key is available in a HyperAI service connector linked to this
+            orchestrator.
+
+        Args:
+            deployment: The pipeline deployment to prepare or run.
+            stack: The stack the pipeline will run on.
+            environment: Environment variables to set in the orchestration
+                environment.
+
+        Raises:
+            RuntimeError: If a step fails.
+        """
+        # Basic Docker Compose definition
+        compose_definition = {"version": "3", "services": {}}
+
+        # Get deployment id
+        deployment_id = deployment.id
+
+        # Set environment
+        os.environ[ENV_ZENML_HYPERAI_RUN_ID] = str(deployment_id)
```

Comment:
I think using the deployment ID here will lead to issues with scheduled pipelines, as there will be multiple runs for the same deployment. (This ID needs to be something that is unique for each run). Maybe docker compose has something like an internal run ID that they use, which we can expose in the environment with some placeholder? 🤔

Reply:
Also, I think you don't actually need to set this in the local environment, only in the containers where the steps get executed.

Reply:
I will test again to make sure I don't, but I added it there because I _did_ get an error related to setting this run ID. If I don't anymore, I will remove it. Thanks!

I will read into possible internal run IDs.

Reply:
Upon removal of that line I'm getting this error:

```
│ /mnt/c/users/cvers/repos/zenml/src/zenml/integrations/hyperai/orchestrators/hyperai_orchestrator │
│ .py:93 in get_orchestrator_run_id                                                                │
│                                                                                                  │
│    90 │   │   try:                                                                               │
│    91 │   │   │   return os.environ[ENV_ZENML_HYPERAI_RUN_ID]                                    │
│    92 │   │   except KeyError:                                                                   │
│ ❱  93 │   │   │   raise RuntimeError(                                                            │
│    94 │   │   │   │   "Unable to read run id from environment variable "                         │
│    95 │   │   │   │   f"{ENV_ZENML_HYPERAI_RUN_ID}."                                             │
│    96 │   │   │   )                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
RuntimeError: Unable to read run id from environment variable ZENML_HYPERAI_ORCHESTRATOR_RUN_ID.
```

I'm using it there because both the Airflow and Local Docker orchestrators have similar `get_orchestrator_run_id` defs.

Any suggestions how to proceed?

Reply:
Also, is `get_orchestrator_run_id` called within the container during a pipeline run? If so, we could perhaps set a unique run ID there via e.g. a timestamp.

Reply:
I think we're close then, because we will then probably run into the issue that two scheduled pipelines are not supported (as they share the same environment variable).

How about changing

```
f"(crontab -l ; echo '{cron_expression} cd {directory_name} && docker compose up -d') | crontab -"
```

into

```
deployment_id = deployment.id
deployment_id_uppercase = deployment_id.upper()
f"(crontab -l ; echo '{cron_expression} cd {directory_name} && export ZENML_HYPERAI_RUN_ID_{deployment_id_uppercase}=\"{deployment_id}_$(date +%s)\" && docker compose up -d') | crontab -"

environment_run_id = f"{ENV_ZENML_HYPERAI_RUN_ID_{deployment_id_uppercase}"
environment[environment_run_id] = str(deployment_id)
```

I believe this means that:

* Each run will have its own unique run ID.
* Multiple scheduled pipelines will never have conflicting runs environment variable wise.

Reply:
That would not work with the implementation of `get_orchestrator_run_id(...)` which checks for a fixed env variable, right?

Reply:
But I don't think we need the `_{deployment_id_uppercase}` suffix in the environment variable, we can just keep that as `ZENML_HYPERAI_RUN_ID` and set it to different values right?

Reply:
I guess if you ssh into the instance, you get a separate shell in which you set the environment variable, which won't interfere with other sessions? At least that was my assumption, not sure it holds true though. As for long-running pipelines: I'm not sure if the environment variables get resolved by `docker-compose up` or once the container starts running (the latter could potentially be an issue). Is there maybe a way to run `docker-compose up -e ZENML_HYPERAI_RUN_ID=...` or something comparable 🤔)

Reply:
You're right! I tested as follows:

```
# m h  dom mon dow   command
* * * * * export VARIABLE_CONTENT="first_job" && echo $VARIABLE_CONTENT > /home/chris/first_job.txt
* * * * * export VARIABLE_CONTENT="second_job" && echo $VARIABLE_CONTENT > /home/chris/second_job.txt
```

The result:

```
(zenmldevelop) chris@MSI:~$ cat first_job.txt
first_job
(zenmldevelop) chris@MSI:~$ cat second_job.txt
second_job
```

So unless you have a different approach that works better, I'll make sure that the crontab entry is created as follows tomorrow - and then test it 😄 

```
f"(crontab -l ; echo '{cron_expression} cd {directory_name} && export ZENML_HYPERAI_RUN_ID=\"{deployment_id}_$(date +%s)\" && docker compose up -d') | crontab -"
```

Reply:
(As for longer-running pipelines: good question... looks like something is available. I'll look into it tomorrow: https://docs.docker.com/compose/environment-variables/set-environment-variables/#set-environment-variables-with-docker-compose-run---env).

Reply:
Current idea:

* Write env var to a pipeline-specific env file in the pipeline folder created by the orchestrator (deterministic .env file location)
* Include it in the Docker Compose file for the services.

That way, each time it's executed, the new env var is set and then passed to the Docker daemon. We can then also test what happens with longer-running pipelines.

Reply:
ENV INTERACTION FOR DIFFERENT PIPELINE RUNS:
To see if two different pipelines can interfere each other when scheduled, I've tested with the following crontab:

12 * * * * cd /mnt/c/users/cvers/repos/composetest/serv1 && echo "SOME_ENV_VAR=1" > .env1 && docker compose up -d
11 * * * * cd /mnt/c/users/cvers/repos/composetest/serv2 && echo "SOME_ENV_VAR=2" > .env2 && docker compose up -d

Both entries start up a sequential series of 3 containers, each of which echos $SOME_ENV_VAR and then sleeps for 120 seconds. Because one `compose up` starts one minute after the other, there are 3 windows of overlap where envvar interaction might take place. If it doesn't, we'd expect 3x1 and 3x2 when inspecting the logs after the whole sequence finishes.

Results align with expectations - when logging the container logs I see 3x1 and 3x2. We can thus be safe to assume that 

root@MSI:/mnt/c/users/cvers/repos/composetest/serv2# docker logs 3c74
Contents: 2

ENV INTERACTION FOR SAME PIPELINE RUN:
This also means that if a pipeline is scheduled twice to run at different intervals, it will work nicely. That is because ZenML will place the pipeline twice in a different folder based on the following run ID, which I just tested is different every single time you run a pipeline (for scheduling):

```
orchestrator_run_id = self.get_orchestrator_run_id()
```

...so there will be no interference when scheduling the same pipeline twice...

```
zenml@ubuntu-4gb-fsn1-2:~/scheduled-pipeline-runs$ ls
1a5501d9-af43-4f98-b0f8-2c3336f689df  4d2514c6-1a4e-4359-9ac2-7907b54b5f88
```

```
zenml@ubuntu-4gb-fsn1-2:~/scheduled-pipeline-runs$ crontab -l
*/5 * * * * cd /home/zenml/scheduled-pipeline-runs/1a5501d9-af43-4f98-b0f8-2c3336f689df && docker compose up -d
*/5 * * * * cd /home/zenml/scheduled-pipeline-runs/4d2514c6-1a4e-4359-9ac2-7907b54b5f88 && docker compose up -d
```

Hence, I'll work on the proposed implementation now :)

Reply:
Implemented in https://github.com/zenml-io/zenml/pull/2372/commits/7bb3405f4b831b591b7c39ecbb984079e0cbc1d6 and tested with both scheduled and nonscheduled pipeline runs. a92dd2195 adds more information to the docs.

With schedule (`schedule = Schedule(cron_expression="*/1 * * * *")`) the pipeline run nicely executes every minute.

Without the schedule the pipeline starts up immediately and finishes nicely.

Reply:
Nice!